rust/libnewsboat/src/stflrichtext.rs

```diff
@@ -0,0 +1,152 @@
+            let Some(tag_start) = tag_start else {
+                break;
+            };
+            if remainder[tag_start..].starts_with('>') {
+                // Keep unmatched '>' (stfl way of encoding a literal '>')
+                pos += tag_start + 1;
+                continue;
+            }
+            let remainder_in_tag = &remainder[tag_start + 1..];
+            let tag_end = opt_first(remainder_in_tag.find('<'), remainder_in_tag.find('>'))
+                .map(|i| i + tag_start + 1);
+            let Some(tag_end) = tag_end else {
+                break;
+            };
+            if remainder[tag_end..].starts_with('<') {
+                // First '<' bracket is unmatched, ignoring it
+                pos += tag_start + 1;
+                continue;
+            }
+            if tag_end - tag_start == 1 {
+                // Convert "<>" into "<" (stfl way of encoding a literal '<')
+                result.remove(pos + tag_end);
+                pos += tag_start + 1;
+                continue;
+            }
+            tags.insert(
+                pos + tag_start,
+                remainder[tag_start..tag_end + 1].to_owned(),
+            );
+            result.replace_range(pos + tag_start..pos + tag_end + 1, "");
+            pos += tag_start;
+        }
+
+        (result, tags)
+    }
+
+    fn merge_style_tag(&mut self, tag: &str, start: usize, end: usize) {
+        if end <= start {
+            return;
+        }
+
+        // Find the latest tag occurring before `end`.
+        // It is important that looping executes in ascending order of location.
+        let mut latest_tag = "</>";
+        for (&pos, tag) in self.style_tags.iter() {
+            if pos > end {
+                break;
+            }
+            latest_tag = tag;
+        }
+        let previous_tag = latest_tag.to_owned();
+        self.style_tags.insert(start, tag.into());
+        self.style_tags.insert(end, previous_tag);
+
+        // Remove any old tags between the start and end marker
+        self.style_tags.retain(|&k, _v| k <= start || k >= end);
+    }
+
+    fn insert_style_tags(&self) -> String {
+        let mut result = self.text.clone();
+        let mut tags = self.style_tags.clone();
+
+        // Expand "<" into "<>" (reverse of what happened in extract_style_tags()
+        let mut pos = 0;
+        while pos < result.len() {
+            let bracket = result[pos..].find('<').map(|i| i + pos);
+            let Some(bracket) = bracket else {
+                break;
+            };
+            pos = bracket + 1;
+            // Add to strings in the `tags` map so we don't have to shift all the positions in that map
+            // (would be necessary if inserting directly into `str`)
+            tags.insert(pos, '>'.into());
+        }
+
+        for (&pos, tag) in tags.iter().rev() {
+            result.insert_str(pos, tag);
+        }
+
+        result
```

Comment:
That will fail about 15 tests (because the last few lines actually add tags into the resulting strings)
I've added a comment to put a bit more emphasis on that point.

If I just replace the while loop with `result.replace("<", "<>")`, there are a few subtle failures (only 2 failing tests) due to the tag insertion positions getting out-of-sync with the result string.

Currently this method is a relatively direct translation from the old code in C++.
To avoid regressions, I prefer to not refactor this too much at this point.

Reply:
> because the last few lines actually add tags into the resulting strings

Oh, right, I missed that.

>  To avoid regressions, I prefer to not refactor this too much at this point.

I totally agree.